File: frontend/src/api/webhooks.ts
import { api } from './client';

export interface Webhook {
  id: string;
  name: string;
  url: string;
  events: string[] | null;
  is_active: boolean;
  workspace_id: string;
  created_at: string;
  updated_at: string;
}

export interface WebhookLog {
  id: string;
  event: string;
  payload: Record<string, unknown> | null;
  response_status: number | null;
  success: boolean;
  webhook_id: string;
  created_at: string;
}

export const webhooksApi = {
  list: (workspaceId: string) =>
    api.get<Webhook[]>(`/workspaces/${workspaceId}/webhooks`),

  create: (workspaceId: string, data: { name: string; url: string; secret?: string; events?: string[] }) =>
    api.post<Webhook>(`/workspaces/${workspaceId}/webhooks`, data),

  update: (workspaceId: string, webhookId: string, data: Partial<Webhook>) =>
    api.put<Webhook>(`/workspaces/${workspaceId}/webhooks/${webhookId}`, data),

  delete: (workspaceId: string, webhookId: string) =>
    api.delete(`/workspaces/${workspaceId}/webhooks/${webhookId}`),

  logs: (workspaceId: string, webhookId: string) =>
    api.get<WebhookLog[]>(`/workspaces/${workspaceId}/webhooks/${webhookId}/logs`),
};

File: backend/app/api/webhooks.py
import uuid

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.models.webhook import Webhook, WebhookLog
from app.models.user import User
from app.schemas.webhook import (
    WebhookCreate,
    WebhookLogResponse,
    WebhookResponse,
    WebhookUpdate,
)
from app.utils.auth import get_current_user

router = APIRouter(
    prefix="/workspaces/{workspace_id}/webhooks",
    tags=["webhooks"],
)


@router.get("", response_model=list[WebhookResponse])
async def list_webhooks(
    workspace_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(Webhook)
        .where(Webhook.workspace_id == workspace_id)
        .order_by(Webhook.created_at.desc())
    )
    return result.scalars().all()


@router.post("", response_model=WebhookResponse, status_code=201)
async def create_webhook(
    workspace_id: uuid.UUID,
    data: WebhookCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    webhook = Webhook(
        name=data.name,
        url=data.url,
        secret=data.secret,
        events=data.events,
        is_active=data.is_active,
        workspace_id=workspace_id,
    )
    db.add(webhook)
    await db.commit()
    await db.refresh(webhook)
    return webhook


@router.put("/{webhook_id}", response_model=WebhookResponse)
async def update_webhook(
    workspace_id: uuid.UUID,
    webhook_id: uuid.UUID,
    data: WebhookUpdate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(Webhook).where(
            Webhook.id == webhook_id,
            Webhook.workspace_id == workspace_id,
        )
    )
    webhook = result.scalar_one_or_none()
    if not webhook:
        raise HTTPException(status_code=404, detail="Webhook not found")
    for k, v in data.model_dump(exclude_unset=True).items():
        setattr(webhook, k, v)
    await db.commit()
    await db.refresh(webhook)
    return webhook


@router.delete("/{webhook_id}", status_code=204)
async def delete_webhook(
    workspace_id: uuid.UUID,
    webhook_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(Webhook).where(
            Webhook.id == webhook_id,
            Webhook.workspace_id == workspace_id,
        )
    )
    webhook = result.scalar_one_or_none()
    if not webhook:
        raise HTTPException(status_code=404, detail="Webhook not found")
    await db.delete(webhook)
    await db.commit()


@router.get("/{webhook_id}/logs", response_model=list[WebhookLogResponse])
async def list_webhook_logs(
    workspace_id: uuid.UUID,
    webhook_id: uuid.UUID,
    limit: int = Query(50, le=200),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(WebhookLog)
        .where(WebhookLog.webhook_id == webhook_id)
        .order_by(WebhookLog.created_at.desc())
        .limit(limit)
    )
    return result.scalars().all()
